### ISSUE TITLE
GNUPG Folder Permission Error When Adding a Public PGP Key

### ISSUE THREAD

**Issue by glenn-sorrentino** [scidsg/hushline#305]
The error is odd - it's a permission issue, but the installation script explicitly grants the folder permission. 

After a successful installation, when I attempt to add a pgp key in settings, I get the following error:
```
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: PermissionError: [Errno 13] Permission denied: '/root/.gnupg'
Mar 28 02:05:27 frontpage-dev gunicorn[18552]:     raise child_exception_type(errno_num, err_msg, err_filename)
Mar 28 02:05:27 frontpage-dev gunicorn[18552]:   File "/usr/lib/python3.11/subprocess.py", line 1901, in _execute_child
Mar 28 02:05:27 frontpage-dev gunicorn[18552]:     self._execute_child(args, executable, preexec_fn, close_fds,
Mar 28 02:05:27 frontpage-dev gunicorn[18552]:   File "/usr/lib/python3.11/subprocess.py", line 1024, in __init__
Mar 28 02:05:27 frontpage-dev gunicorn[18552]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Mar 28 02:05:27 frontpage-dev gunicorn[18552]:     return subprocess.Popen(cmd, shell=expand_shell, stdin=subprocess.PIPE,
Mar 28 02:05:27 frontpage-dev gunicorn[18552]:   File "/var/www/html/test.ourdemo.app/venv/lib/python3.11/site-packages/gnupg/_meta.py", line 620, in _open_subprocess
Mar 28 02:05:27 frontpage-dev gunicorn[18552]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Mar 28 02:05:27 frontpage-dev gunicorn[18552]:     proc = self._open_subprocess(["--list-config", "--with-colons"])
Mar 28 02:05:27 frontpage-dev gunicorn[18552]:   File "/var/www/html/test.ourdemo.app/venv/lib/python3.11/site-packages/gnupg/_meta.py", line 506, in _check_sane_and_get_gpg_version
Mar 28 02:05:27 frontpage-dev gunicorn[18552]:     self._check_sane_and_get_gpg_version()
Mar 28 02:05:27 frontpage-dev gunicorn[18552]:   File "/var/www/html/test.ourdemo.app/venv/lib/python3.11/site-packages/gnupg/_meta.py", line 242, in __init__
Mar 28 02:05:27 frontpage-dev gunicorn[18552]:     super(GPG, self).__init__(
Mar 28 02:05:27 frontpage-dev gunicorn[18552]:   File "/var/www/html/test.ourdemo.app/venv/lib/python3.11/site-packages/gnupg/gnupg.py", line 117, in __init__
Mar 28 02:05:27 frontpage-dev gunicorn[18552]:           ^^^^^^^^^^^^^^^^^^^
Mar 28 02:05:27 frontpage-dev gunicorn[18552]:     gpg = gnupg.GPG(gpg_home)
Mar 28 02:05:27 frontpage-dev gunicorn[18552]:   File "/var/www/html/test.ourdemo.app/hushline/crypto.py", line 30, in is_valid_pgp_key
Mar 28 02:05:27 frontpage-dev gunicorn[18552]:          ^^^^^^^^^^^^^^^^^^^^^^^^^
Mar 28 02:05:27 frontpage-dev gunicorn[18552]:     elif is_valid_pgp_key(pgp_key):
Mar 28 02:05:27 frontpage-dev gunicorn[18552]:   File "/var/www/html/test.ourdemo.app/hushline/settings.py", line 435, in update_pgp_key
Mar 28 02:05:27 frontpage-dev gunicorn[18552]:            ^^^^^^^^^^^^^^^^^^
Mar 28 02:05:27 frontpage-dev gunicorn[18552]:     return f(*args, **kwargs)
Mar 28 02:05:27 frontpage-dev gunicorn[18552]:   File "/var/www/html/test.ourdemo.app/hushline/utils.py", line 18, in decorated_function
Mar 28 02:05:27 frontpage-dev gunicorn[18552]:                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Mar 28 02:05:27 frontpage-dev gunicorn[18552]:     return cast(R, flask.current_app.ensure_sync(obj)(*a, **k))
Mar 28 02:05:27 frontpage-dev gunicorn[18552]:   File "/var/www/html/test.ourdemo.app/venv/lib/python3.11/site-packages/flask_limiter/extension.py", line 1299, in __inner
Mar 28 02:05:27 frontpage-dev gunicorn[18552]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Mar 28 02:05:27 frontpage-dev gunicorn[18552]:     return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]
Mar 28 02:05:27 frontpage-dev gunicorn[18552]:   File "/var/www/html/test.ourdemo.app/venv/lib/python3.11/site-packages/flask/app.py", line 855, in dispatch_request
Mar 28 02:05:27 frontpage-dev gunicorn[18552]:          ^^^^^^^^^^^^^^^^^^^^^^^
Mar 28 02:05:27 frontpage-dev gunicorn[18552]:     rv = self.dispatch_request()
Mar 28 02:05:27 frontpage-dev gunicorn[18552]:   File "/var/www/html/test.ourdemo.app/venv/lib/python3.11/site-packages/flask/app.py", line 870, in full_dispatch_request
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: Traceback (most recent call last):
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: 2024-03-28 02:05:27,043:ERROR:Error: [Errno 13] Permission denied: '/root/.gnupg'
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: 2024-03-28 02:05:27,039:INFO:Setting homedir to '/root/.config/python-gnupg'
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: 2024-03-28 02:05:27,038:INFO:Creating directory: /root/.config/python-gnupg
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: -----END PGP PUBLIC KEY BLOCK-----
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: =t34E
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: 2b+bArPU6cK9
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: pC7bigbPBjNSf3zGFJAdRU9OuBXMSjEFZWLqiJEp+m+b8MQ8PdpGOY7Qg/fU
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: 4OryjRafWE+NuXzUOvic1Q61rmyg8zvGMAySYqxhE0EaKm69Zm/CDsLnUYtn
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: uYi/Mg1IynlJtRadm2S+eMEjyov6Im0RRBUSs0i63iUD/rdIC+FBqDtwaaM4
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: X+kxMuZorpDT92L7Qm6RdZZzxYxxcDAw4iDuP+CDw0OkyVrGjsA2mhtwkn+e
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: hpphonHRb2BQrFl1RE3zX9qe1oQQ7MexUTVB//YOEjxb9OkVJv9NBfboE6BG
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: ig3uGqLvOE8H2Zm02tv8+d2KM1r3rVKUD3sT545PC/1rf30ON/mVk733BYJd
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: D7MA6/cCa88Dxff5S5AusLqrI7D6Z6D3nv4OIwICanyiOHKBkam2kqCw5jA5
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: YtplCBLbpo/ElXviQqzw2RGAsmhX6BYTFD6WJdIDNYeDAdTroQLJd/wflbA/
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: 61BMCcObEMx7x+phZmYL/zcyJ3F9rbZTb0no5DBx7CxJfeZVRX9DJsU3Xdl7
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: MKveiMOA/SPT4zM5+N3hvzVNML9hfHAsVmJua2bYVwcNW0+4Up2s0y83KgCf
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: tHFxHcC80LHdKAaL3ARMpNQr+bkwzxRK3t4fingOlsghqUMuZlyhnURt8hJF
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: aVBXyD4CmwwWIQSGM67QIwMOIsmtmxzgUjFpUFfIPgAAGmAP/36KePu7foyZ
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: +41t4OBwkYWLAcn7208cHngYEM8AEQEAAcLBdgQYAQgAKgWCZTC1rQmQ4FIx
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: 44F5SxlxOwP1Y8MeU/GGyKshpsDCm+kPm8shOYMaWsdiMbJ05qo/ksAcklTB
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: BbhYkauHTCb0HDpuZoR3WlILimDIck3+VpIRAbs0VarUq7Kxf/Gn/evgDzFl
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: vEOWT4GQf1R8zfOmPdE7ygRKJJUUjrBLAjVwTLMh1uK19k7jPKqA4XPo0ZRm
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: otE1Z+vJXOCCmLP3y94AlH1pA77cIpzddi3Wzrs1l/15gsqjfwRbZ0wUTA7o
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: jEkYScT+tSXu1OPufQtQxaNxTqc1mzpkAfGsIleuaiulmz6R6uTzbClyPeN/
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: 6lYPNrNoRPEsmceTopIX5bH3eIeOzCe3/NdEU0FseJalBX+iGO22ago1ESNB
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: k4WAoTzMkqDDDrCApVGtfzuGYkoTS7pOZKVjzq0MrqEkwbPLQ6nzDTl81Q+u
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: jQm0VcvCw6TQMDV5McxXb4vcfSl40fDaNdt4dEd5nYE43WF/mC50vm7h5Awg
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: ACtrSMR5P5xNOa4Pi0hyzP7ZhBqGRz5/WC4+jeF9qOdqaus2YV2pMBx1xAV9
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: sOuCn2hnJJiLJGhhsyPl/vYAfxVn0Eo5ZzyEjVkoFNOiuDGpzYZ4RBKRWru9
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: ARAAsX8HUuX5sxXF0R+2IjjcTsyRumQ0LubD3W/zGSB2fUT2zV+VGgEVdukC
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: i/9J0oZp1NB0MQwu+WgICiAqb03QSdBS9yEMl59a5Ec89ucZ2s7BTQRlMLWt
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: RGSZvQQsoufv0g9iUjESjLWSsX6iBSW3AI5I9kB8E3LUctIFwW14PSDqv22m
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: +HYVJ1eUwqLULmP09Id8An75GSbho92S1eIrVg5FqCnaGlqi8q8ZWJyCN/K7
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: B07CMl5Xdmr9/gWI3Wp4lfCzypxfeGIZ+wEfdPDHaCph0z2jgxcsAtBabLwo
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: DNDgcYwV8ggBMb9BJvN061W1zOwlOaAn2OP5bzWfG1q0oIIIo1YkLFfZmlLU
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: 8bd4UBf2W4GqiltMxQX4ErbaNlT7sr+37PNGqLDShn670zx4+IayfyhcBYMh
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: 8oPswYHhhCZGfWrbyJ1wcoNNw4DxZpd+V2JQNhzs4W69fP71mWodpvmH3Sjj
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: /WwTrmVmv9V8BIAdbKA4lXzufvD1no8IeyIHhgC2LU8d7M3qCdhjMuT5xdPa
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: H/PygHqvIcNd8dkwWy/FZunleIxPl0DO5eBTifx5dya8G6pKEZ7qydHnOQD5
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: ECe/DKwC8HlqIfde7Os1qDqebKGMpHjikMDJOdFmdFszORJTW4MaBhP3AIM0
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: NZNrN4A62aRQMHXvQ1arXWSuVuafjHIEGTmh/WiRIcCj1gxEeKYV0wqxiI9T
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: DiLJrZsc4FIxaVBXyD4AALFiD/9UW0u/x0GGZ779hbLxQUmLkhthzW+CTots
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: MLWtBAsJBwgJkOBSMWlQV8g+AxUICgQWAAIBAhkBApsDAh4BFiEEhjOu0CMD
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: RGVzaWduIFRpcCBMaW5lIDx0aXBzQHNjaWRzZy5vcmc+wsGKBBABCAA+BYJl
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: +MBvcKsq9pLYZvKGGO90mAmjdzJy8agH/herRQARAQABzStTY2llbmNlICYg
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: bbLzMYNngEHcuWt69gdRGR7H/yRdmWmzAzoxWb0vT6UZtuxX1Bkyn9i6bbKE
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: d1jVyN109ZBqBuN71GalUFUng0d2wMEDQlhRxPlh3VasmQS6hc7iIotSXLbe
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: JI+pSbZlO/5YWCELm7lnsrW5YmLeYhqLaaH3O2AD2mG8u2TNNE0hMXMXxjP9
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: XWlP4FAhDVB0cuvDOmgCzQGDohDcItuyXyYIG3q3UivxD9LfjNyRQudQokq2
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: WQpBbr36DQ0sJESbwUndBDO21dcSObElEgtQPttQ81F6sh/aOuNRNhojAJCS
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: MCgC1xv50VwIxwKOhE1QYEIflGZE1kQ0xHEdo2h1Zww0Uo6hrOvWpy/Lba98
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: t8aLdTtfHel4zRoyPGKO/89gNaKje+KFDfpVCJvMWfHCH7pW/b+yivfowqX5
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: aQ+1OmbfvHYEd5EFN/ft7WKxhn5WcgHYvDTyWfnFNvBEAii/x9FQf9m7iVHm
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: HIrj5vFJ1CjKDsJ0tvXO1ZSLpbrSiVV9EoCU/05u6k5KyoYUAMgYElLUVigq
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: f10nTJBTqpR7RQU3ah4P9iaYBBhUOpksB/un6PL6F71N/Qx5o5aUvs5q1ea4
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: xsFNBGUwta0BEADCJuMJgBhFQXm1PAixuH8f3amFHqk+uyoDHLcM9LuN1fPM
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: 
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: Comment: https://mailvelope.com
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: Version: Mailvelope v5.1.1
Mar 28 02:05:27 frontpage-dev gunicorn[18552]: 2024-03-28 02:05:27,037:DEBUG:Attempting to import key: -----BEGIN PGP PUBLIC KEY BLOCK-----
Mar 28 02:04:25 frontpage-dev gunicorn[18552]: [2024-03-28 02:04:25 +0000] [18552] [INFO] Booting worker with pid: 18552
Mar 28 02:04:25 frontpage-dev gunicorn[18551]: [2024-03-28 02:04:25 +0000] [18551] [INFO] Booting worker with pid: 18551
Mar 28 02:04:25 frontpage-dev gunicorn[18537]: [2024-03-28 02:04:25 +0000] [18537] [INFO] Using worker: sync
Mar 28 02:04:25 frontpage-dev gunicorn[18537]: [2024-03-28 02:04:25 +0000] [18537] [INFO] Listening at: unix:/var/www/html/test.ourdemo.app/hushline-hosted.sock (18537)
Mar 28 02:04:25 frontpage-dev gunicorn[18537]: [2024-03-28 02:04:25 +0000] [18537] [INFO] Starting gunicorn 20.1.0
Mar 28 02:04:24 frontpage-dev systemd[1]: Started hushline-hosted.service - Gunicorn instance to serve the Hushline Flask app.
Mar 28 02:04:24 frontpage-dev systemd[1]: hushline-hosted.service: Consumed 2.428s CPU time.
Mar 28 02:04:24 frontpage-dev systemd[1]: Stopped hushline-hosted.service - Gunicorn instance to serve the Hushline Flask app.
Mar 28 02:04:24 frontpage-dev systemd[1]: hushline-hosted.service: Deactivated successfully.
Mar 28 02:04:24 frontpage-dev gunicorn[18505]: [2024-03-28 02:04:24 +0000] [18505] [INFO] Shutting down: Master
Mar 28 02:04:23 frontpage-dev gunicorn[18529]: [2024-03-28 02:04:23 +0000] [18529] [INFO] Worker exiting (pid: 18529)
```

It's odd, because we're setting the permissions in install.sh:
```
mkdir -p ~/.gnupg
chmod 700 ~/.gnupg

# Install dependencies
python3 -m venv venv
source venv/bin/activate
curl -sSL https://install.python-poetry.org | python3 -
export PATH="/root/.local/bin:$PATH"
echo 'export PATH="/root/.local/bin:$PATH"' >> ~/.bashrc
poetry lock
poetry install
```

We're explicity setting the folder location in crypto.py:
```
import os

import gnupg
from cryptography.fernet import Fernet
from flask import current_app

encryption_key = os.environ.get("ENCRYPTION_KEY")
if encryption_key is None:
    raise ValueError("Encryption key not found. Please check your .env file.")

fernet = Fernet(encryption_key)

gpg_home = os.path.expanduser("~/.gnupg")
```

Manual verification of the folder shows that permissions are correct:
```
root@frontpage-dev:~# ls -ld /root/.gnupg
drwx------ 2 root root 4096 Mar 28 02:03 /root/.gnupg
```

Checking to see if GPG can access and modify the directory shows it can without any issues:
```
root@frontpage-dev:~# gpg --list-keys --homedir /root/.gnupg
gpg: keybox '/root/.gnupg/pubring.kbx' created
gpg: /root/.gnupg/trustdb.gpg: trustdb created
root@frontpage-dev:~# 
```

**Comment by brassy-endomorph** [comment]
This might have to do with the fact that the systemd unit file is running as `root:www-data`. I set the unit file to `root:root` for which is hardly a security loss for the moment.

**Comment by glenn-sorrentino** [comment]
That gave a bad gateway. Changing the group back to `www-data` resolved that issue, but adding a pgp key is still causing the app to crash. 

I went at it from another angle - since this was related to `gnupg` and how it handles files, I thought to try changing algos. I upgraded to `pysequoia` and got it working. It's in the `sequoia` branch.

**Comment by glenn-sorrentino** [comment]
Closed with #308
